synapse/axon.py

```diff
@@ -365,20 +444,23 @@ async def metrics(self):
     async def save(self, sha256, genr):
 
         self._reqBelowLimit()
-        byts = self.axonslab.get(sha256, db=self.sizes)
-        if byts is not None:
-            return int.from_bytes(byts, 'big')
 
-        size = await self._saveFileGenr(sha256, genr)
+        async with self.holdHashLock(sha256):
 
-        self._addSyncItem((sha256, size))
+            byts = self.axonslab.get(sha256, db=self.sizes)
+            if byts is not None:
+                return int.from_bytes(byts, 'big')
```

Comment:
we should dig up here and add test coverage

Reply:
Added a test.  In general this block isn't hit since Upload.save() checks if the bytes exist before calling Axon.save(), but the scenario I covered is two concurrent uploads of the same file, where the second is waiting for the lock in Axon.save() and then the file exists once it comes back.